File: src/main/java/club/sk1er/patcher/tweaker/asm/GuiMultiplayerTransformer.java
package club.sk1er.patcher.tweaker.asm;

import club.sk1er.patcher.tweaker.transform.PatcherTransformer;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.*;

public class GuiMultiplayerTransformer implements PatcherTransformer {
    @Override
    public String[] getClassName() {
        return new String[]{"net.minecraft.client.gui.GuiMultiplayer"};
    }

    @Override
    public void transform(ClassNode classNode, String name) {
        for (MethodNode method : classNode.methods) {
            String methodName = mapMethodName(classNode, method);

            if (methodName.equals("keyTyped") || methodName.equals("func_73869_a")) {
                method.instructions.insertBefore(method.instructions.getFirst(), selectServer());
                break;
            }
        }
    }

    private InsnList selectServer() {
        InsnList list = new InsnList();
        list.add(new VarInsnNode(Opcodes.ALOAD, 0));
        list.add(new MethodInsnNode(Opcodes.INVOKESTATIC, "club/sk1er/patcher/hooks/GuiMultiplayerHook", "keyTyped", "(Lnet/minecraft/client/gui/GuiMultiplayer;)V", false));
        return list;
    }
}

File: src/main/kotlin/club/sk1er/patcher/asm/GuiMultiplayerTransformer.kt
package club.sk1er.patcher.asm

import club.sk1er.patcher.asm.utils.getInstructions
import club.sk1er.patcher.hooks.GuiMultiplayerHook
import club.sk1er.patcher.tweaker.transform.PatcherTransformer
import org.objectweb.asm.tree.ClassNode

class GuiMultiplayerTransformer : PatcherTransformer {
    override fun getClassName() = arrayOf("net.minecraft.client.gui.GuiMultiplayer")

    override fun transform(classNode: ClassNode, name: String) {
        classNode.methods.first {
            val methodName = mapMethodName(classNode, it)
            methodName == "keyTyped" || methodName == "func_73869_a"
        }?.apply {
            instructions.insertBefore(instructions.first, getInstructions {
                of(GuiMultiplayerHook::keyTyped)
                target(this@apply)
                before(this@apply.instructions.first)
                param(0)
            })
        }
    }
}